proposals/0089-programify-feature-gate-program.md

```diff
@@ -0,0 +1,149 @@
+program could be changed only once, eliminating this duplication of work.
+
+## New Terminology
+
+- Feature Gate program: The Core BPF program that all feature accounts will be
+  assigned to, with address `Feature111111111111111111111111111111111111`.
+- "Revoke" or "revoke pending activation": The act of reallocating a feature
+  account's data to zero, assigning it to the System Program, and defunding its
+  lamports balance - effectively removing it from the runtime's recognized set
+  of pending feature activations.
+
+## Detailed Design
+
+The Feature Gate program shall be a Core BPF program whose upgrade authority
+will be controlled by a multi-sig authority, with keyholders from each validator
+client implementation. This includes Solana Labs, Jito, and Jump. In the future,
+this can be expanded to include new clients such as Syndica.
+
+The program shall initially be designed to support one instruction:
+`RevokePendingActivation`. Any other instructions or functionality this program
+may support in the future is outside the scope of this SIMD.
+
+As mentioned above under "New Terminology", when this instruction is invoked by
+a feature key-holder, the program will reallocate the account to zero, assign it
+back to the System Program, and defund its lamports balance. As a result, the
+runtime will no longer recognize this feature as pending, since it will no
+longer be owned by `Feature111111111111111111111111111111111111`.
+
+Consider the instruction as it may appear in the Feature Gate program:
+
+```rust
+pub enum FeatureGateInstruction {
+    /// Revoke a pending feature activation.
+    ///
+    /// A "pending" feature activation is a feature account that has been
+    /// allocated and assigned, but hasn't yet been updated by the runtime
+    /// with an `activation_slot`.
+    ///
+    /// Features that _have_ been activated by the runtime cannot be revoked.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[w+s]`    Feature account
+    ///   1. `[w]`      Destination (for rent lamports)
+    RevokePendingActivation,
+}
+```
+
+The non-existent program at `Feature111111111111111111111111111111111111` can be
+considered analogous to a no-op native program. Thus, the official processes
+outlined in
+[SIMD 0088](https://github.com/solana-foundation/solana-improvement-documents/pull/88)
+for migrating a native program to Core BPF can be used to enable this new program.
+
+Consider the following steps to activate Feature Gate:
+
+1. Migrate the native no-op program at
+   `Feature111111111111111111111111111111111111` to a Core BPF no-op.
+2. Upgrade the Core BPF no-op to add the `RevokePendingActivation` instruction.
+
+Executing these two steps would effectively activate Feature Gate without any
+changes to existing processes.
+
+As mentioned previously, the upgrade in step 2 would be conducted by the Feature
+Gate program's multi-sig upgrade authority. 
```

Comment:
```suggestion
The official processes
outlined in
[SIMD 0088](https://github.com/solana-foundation/solana-improvement-documents/pull/88)
for migrating a native program to Core BPF will be used to enable this new program, with the addition of the Revoke instruction as a separate BPF rogram-upgrade step.

1. Migrate the native no-op program at
   `Feature111111111111111111111111111111111111` to a Core BPF no-op, with the new program's upgrade authority set to the Feature Gate multisig
2. Upgrade the Core BPF no-op to add the `RevokePendingActivation` instruction using the new multisig authorization

Because the only change to the program is the Revoke addition, these steps will enable BPF Feature Gate without any
changes to existing feature-activation processes.
```